runtime/extern.c

```diff
@@ -786,7 +805,10 @@ static void extern_rec(struct caml_extern_state* s, value v)
     /* Check if object already seen */
     if (! (s->extern_flags & NO_SHARING)) {
       if (extern_lookup_position(s, v, &pos, &h)) {
-        extern_shared_reference(s, s->obj_counter - pos);
+        /* #4056: using absolute references for shared objects improves
+           compressibility. */
+        uintnat d = s->extern_flags & COMPRESSED ? pos : s->obj_counter - pos;
+        extern_shared_reference(s, d);
```

Comment:
I have mixed feeling about using conditional logic here. It makes the code more complex, and we lose the property that the only difference between both payloads is compression. Are we sure that it is worth it? (Are the space saving in the uncompressed case substantial enough to justify relative offsets, and the space saving in the compressed case substantial enough to justify absolute offsets?)

Reply:
As mentioned in 6f0b636c06, switching to absolute references reduced the size of a big .cmt file (probably camlinternalFormat.cmt) by 15% *on top of the savings obtained by zlib compression*.  I haven't re-done the measurements since switching to zstd.  But it's clear that the insight in #4056 is correct and that the compressed format must use absolute references.

On the other hand, you don't change the existing uncompressed formats lightly.  Switching to absolute references there would mean new magic numbers, inability to read persistent data marshaled with older OCaml versions, and perhaps some "fun" in the bootstrapping process.  Or, if you want backward compatibility, you'd need to support both absolute and relative references, with different magic numbers, and the "conditional logic" you dislike would be there again.  I don't even want to know which of relative or absolute references perform better without compression.